### DISCUSSION TITLE
Client Vs Server Components

### DISCUSSION THREAD

**Discussion by anish-fusemachines** [vercel/next.js#51914] [Help] [answered]
### Summary

In nextjs 13 we can specify client side using "use client" . In my homepage i'm using different plugins like carousel , TwitterTimelineEmbed but they can only run in client silde so i have to specify "use client" on top. I think that is creating my whole homepage(page.js) client side.
1. what should i do if i don't want to make my whole page as client rendering?
2. As i know with server side rendering when we  "view page source" we can see all html generated but even after i specify "use client" on top i still can see whole html. is it normal?

### Additional information

```js
// "use client";

import Link from "next/link";

import { TwitterTimelineEmbed } from "react-twitter-embed";

import TrustedBy from "./components/TrustedBy";
import TestimonialSlider from "./components/TestimonialSlider";
import ChipListSection from "./components/ChipListSection";
import CapabilitiesSlider from "./components/CapabilitiesSlider";
```

**Comment by icyJoseph** [comment, marked as answer]
Hi,

There's a widespread misconception about client components, and that is, that they only render client side.

That's not correct. Client Components are the components we have used up till now. They render during SSR and are hydrated Client Side. 

What Client Component means, is that there'll be a bundle of JavaScript that drives this component, this bundle is sent to the client. 

A Server Component's JavaScript is strictly run server side and it is never sent to the the client.

React knows how to weave these two together, to produce the final HTML. 



From https://nextjs.org/docs/getting-started/react-essentials#client-components:

> Client Components enable you to add client-side interactivity to your application. In Next.js, they are [pre-rendered](https://nextjs.org/docs/app/building-your-application/rendering#component-level-client-and-server-rendering) on the server and [hydrated](https://nextjs.org/docs/app/building-your-application/rendering#component-level-client-and-server-rendering) on the client. You can think of Client Components as how components in the [Pages Router](https://nextjs.org/docs/pages) have always worked.

Also from the Next.js docs, https://nextjs.org/docs/getting-started/react-essentials:

> - Components in the Server Component module graph are guaranteed to be only rendered on the server.
> - Components in the Client Component module graph are primarily rendered on the client, but with Next.js, they can also be pre-rendered on the server and hydrated on the client.

Last but not least a word from Dan: https://github.com/reactwg/server-components/discussions/4

So what you can actually do, is think about what parts of your React tree can be moved to a Server Component, and which need to be interactive, so that you can reduce the amount of JS needed to drive your app. Sort of, Carve Out the static parts of your app, so that they don't need to be hydrated.

For example, if you have a page that uses state and event handlers, you might benefit from refactoring so that JSX that's just wrapping  the actual business elements of your component, is just a server component. Simple example:

```jsx
import { useState } from 'react';

const Counter = () => {
  const [count, setCount] = useState(0);
  return (
    <main>
      <header>Here is the count!</header>
      <section>
        <p>{count}</p>
        <button onClick={() => setCount((x) => x + 1)}>+1</button>
      </section>
      <footer>That was the count!</footer>
    </main>
  );
};
```

Could be refactored into:

```jsx
// server component by default, since it doesn't use the "use client" directive
import { Counter } from './Counter';

const CounterArea = () => {
  return (
    <main>
      <header>Here is the count!</header>
      <section>
           <Counter />
      </section>
      <footer>That was the count!</footer>
    </main>
  );
};
```

And now, IN ANOTHER FILE, you define:

```jsx
"use client";

import { useState } from "react";

export const Counter = () => {
  const [count, setCount] = useState(0);
  return (
    <>
      <p>{count}</p>
      <button onClick={() => setCount((x) => x + 1)}>+1</button>
    </>
  );
};
```

It has to be another file, because the module is the Network boundary.

Of course as the React tree complexity grows, and where common interactivity patterns spread through out the tree, it might be harder to carve out some of the wrapping HTML.

Still, some data can be fetched server side, and you get access to Node APIS and what not.  

And what you get out of this, is that there'll be no JavaScript for the header, footer, and main elements, they don't need to be turned into React Elements (plain objects describing the UI), they are just rendered to HTML and React knows how to weave the counter stuff into them.

**Reply by leerob** [reply]
Appreciate this thoughtful response!
